Convert ltac2 chapter to use prodn, update syntax (#12085)
coq/coq: doc/sphinx/proof-engine/ltac2.rst

implementation of Ltac, described in the previous chapter, will be referred to
as Ltac1.

Current limitations include:

- There are a number of tactics that are not yet supported in Ltac2 because
  the interface OCaml and/or Ltac2 notations haven't been written.  See
  :ref:`defining_tactics`.

- Missing usability features such as:

  - Printing functions are limited and awkward to use.  Only a few data types are
    printable.
  - Deep pattern matching and matching on tuples don't work.
  - If statements on Ltac2 boolean values

- Missing low-level primitives that may be needed for writing automation.  For example,
  there's no way to get the number of constructors of an inductive type so it's
  not possible to write automation that's generic over inductive types.  It's not

>   possible to build terms with casts through the low-level API because the cast type

Reviewer: This is not quite true, you can match on `'I <: True` to extract the cast type.  But it's rather painful.

Author: If you can give me a short example I could add that to the chapter as a workaround.

Reviewer: I'm not in front of a computer right now, so can't test this, but something like:
```coq
Ltac2 get_vm_cast () :=
  match Constr.Unsafe.kind '(I <: True) with
  | Constr.Unsafe.Cast _ cst _ => cst
  | _ => Control.throw Not_found
  end.
```

Reviewer: You need to do something like `let cst := get_vm_cast () in ...`. The message means that you're ignoring the return value of the tactic.

Reviewer: This text needs to be updated now that you have the cast example, right?

Author: I didn't understand the type case example 100%.  IIUC it shows how to extract info from a cast; it's not a complete example of "A convenient way to **build terms** with casts".  Perhaps that should be described differently.

We don't actually define "cast" in the documentation.  It converts an expression from one type to a compatible more-specific type?  Processed during type checking, not at runtime?

So I don't see how to "get the number of constructors of an inductive type".

After tweaking this, are there other good examples of "Missing low-level primitives that may be needed for writing automation. " or should I drop that bullet point?

Author: I added your casting example, though I don't know casts Coq well enough to understand how it works.  In some later PR we'll add a better description of casts.

FWIW, the Ltac2 chapter does say "Unlike Ltac1, there are no type casts at runtime".  That may have been in the back of my mind.

Your other points in this comment are covered now.

implementation of Ltac, described in the previous chapter, will be referred to
as Ltac1.

Current limitations include:

- There are a number of tactics that are not yet supported in Ltac2 because
  the interface OCaml and/or Ltac2 notations haven't been written.  See
  :ref:`defining_tactics`.

- Missing usability features such as:

  - Printing functions are limited and awkward to use.  Only a few data types are
    printable.
  - Deep pattern matching and matching on tuples don't work.
  - If statements on Ltac2 boolean values

>   - A convenient way to build terms with casts through the low-level API because the
>     cast type is opaque.  You can use a variation of this code as a workaround:
> 
>     .. coqtop:: none
> 
>        From Ltac2 Require Import Ltac2.
> 
>     .. coqtop:: in
> 
>        Ltac2 get_vm_cast () :=
>          match Constr.Unsafe.kind '(I <: True) with
>          | Constr.Unsafe.Cast _ cst _ => cst
>          | _ => Control.throw Not_found
>          end.
> 
> - Missing low-level primitives that are convenient for writing automation, such as:
> 
>   - An easy way to get the number of constructors of an inductive type.
>     Currently only way to do this is to destruct a variable of the inductive type
>     and count the number of goals that result.
> 
>   - Building terms with casts currently requires an awkward construction like the
>     following, which also incurs extra overhead to repeat typechecking for each
>     call to `get_vm_cast`: `Constr.Unsafe.make (Constr.Unsafe.Cast 'I (get_vm_cast ()) 'True)`

Reviewer: I still think the bullet points should be combined, e.g., like this:
```suggestion
  - A convenient way to build terms with casts through the low-level API. Because the
    cast type is opaque, building terms with casts currently requires an awkward construction like the
    following, which also incurs extra overhead to repeat typechecking for each
    call to `get_vm_cast`: `Constr.Unsafe.make (Constr.Unsafe.Cast 'I (get_vm_cast ()) 'True)`
    with:

    .. coqtop:: none

       From Ltac2 Require Import Ltac2.

    .. coqtop:: in

       Ltac2 get_vm_cast () :=
         match Constr.Unsafe.kind '(I <: True) with
         | Constr.Unsafe.Cast _ cst _ => cst
         | _ => Control.throw Not_found
         end.

- Missing low-level primitives that are convenient for writing automation, such as:

  - An easy way to get the number of constructors of an inductive type.
    Currently only way to do this is to destruct a variable of the inductive type
    and count the number of goals that result.
```

Feel free to merge with or without this, though.

Author: Got, it thanks.

implementation of Ltac, described in the previous chapter, will be referred to
as Ltac1.

Current limitations include:

- There are a number of tactics that are not yet supported in Ltac2 because
  the interface OCaml and/or Ltac2 notations haven't been written.  See
  :ref:`defining_tactics`.

- Missing usability features such as:

  - Printing functions are limited and awkward to use.  Only a few data types are
    printable.
  - Deep pattern matching and matching on tuples don't work.
  - If statements on Ltac2 boolean values

>   - A convenient way to build terms with casts is through the low-level API because the

Reviewer: ```suggestion
  - A convenient way to build terms with casts through the low-level API because the
```
The missing feature is the convenient way to build the terms.

    cast type is opaque.  You can use a variation of this code as a workaround:

    .. coqtop:: none

       From Ltac2 Require Import Ltac2.

    .. coqtop:: in

       Ltac2 get_vm_cast () :=
         match Constr.Unsafe.kind '(I <: True) with
         | Constr.Unsafe.Cast _ cst _ => cst
         | _ => Control.throw Not_found
         end.

- Missing low-level primitives that are convenient for writing automation, such as:

  - An easy way to get the number of constructors of an inductive type.
    Currently only way to do this is to destruct a variable of the inductive type
    count the number of goals that result.

  - Building terms with casts currently requires an awkward construction like the
    following, which also incurs extra overhead to repeat typechecking for each
    call to `get_vm_cast`.

    .. coqdoc::

>        Constr.Unsafe.make (Constr.Unsafe.Cast 'I (get_vm_cast ()) 'True)

Reviewer: This should probably be moved up and fused with the bullet point above about casts, since there's really only one issue here.

         | _ => Control.throw Not_found
         end.

- Missing low-level primitives that are convenient for writing automation, such as:

  - An easy way to get the number of constructors of an inductive type.
    Currently only way to do this is to destruct a variable of the inductive type

>     count the number of goals that result.

Reviewer: ```suggestion
    and count the number of goals that result.
```

The non-terminal :production:`lident` designates identifiers starting with a
lowercase.

.. insertprodn tac2type5 typ_param

.. prodn::
   tac2type5 ::= @tac2type2 -> @tac2type5
   | @tac2type2
   tac2type2 ::= @tac2type1 * {+* @tac2type1 }
   | @tac2type1
   tac2type1 ::= @tac2type0 @qualid
   | @tac2type0
   tac2type0 ::= @typ_param
   | _
   | @qualid
   | ( {+, @tac2type5 } ) @qualid
   | ( @tac2type5 )

>    typ_param ::= ' @ident

Reviewer 2: Not sure whether it's a bug or not. It doesn't seem problematic from the point of view of syntactic clashes, but I wouldn't bet my life on this.

One can define new types with the following commands.

.. cmd:: Ltac2 Type {? rec } @tac2typ_def {* with @tac2typ_def }
   :name: Ltac2 Type

   .. insertprodn tac2typ_def tac2rec_field

   .. prodn::
      tac2typ_def ::= {? @tac2typ_prm } @qualid {? {| := | ::= } @tac2typ_knd }
      tac2typ_prm ::= @typ_param
      | ( {+, @typ_param } )
      tac2typ_knd ::= @tac2type5
      | [ {? {? %| } {+| @tac2alg_constructor } } ]
      | [ .. ]
      | %{ {? {+; @tac2rec_field } {? ; } } %}
      tac2alg_constructor ::= @ident
      | @ident ( {*, @tac2type5 } )
      tac2rec_field ::= {? mutable } @ident : @tac2type5

   :n:`:=`

>      Defines a type with a manifest

Reviewer 2: I think it's the official nomenclature in OCaml, at least in the implementation.

>      Defines a type with a manifest

Reviewer: I think "manifest" definitely needs to be defined here.  Maybe "defines a type with an explicit set of constructors"?  It may be official in the OCaml world, but it's not standard in the Coq world.

   :n:`::=`

>      Defines an open variant type, a special kind of variant type whose constructors are not

Reviewer 3: IIUC this should not be "defines an open variant type" but rather "extends an existing open variant type". The way to define an open variant type would be to use `:= [ .. ]`. Example:

```coq
From Ltac2 Require Import Ltac2.
Ltac2 Type foo := [ .. ].
Ltac2 Type foo ::= [ Bla ].
```

Whereas if you use, `::=` to start with, you get an "unbound type" error.

   Without :n:`{| := | ::= }`
     Defines an abstract type for use representing data from OCaml.  Not for
     end users.

   Each production of :token:`tac2typ_knd` defines one of four possible kinds
   of definitions, respectively: alias, variant, open variant and record types.

   .. cleanup
       .. productionlist:: coq
          ltac2_typedef    : `ltac2_type`
                           : [ `ltac2_constructordef` | ... | `ltac2_constructordef` ]
                           : { `ltac2_fielddef` ; ... ; `ltac2_fielddef` }
                           : [ .. ]
          ltac2_constructordef   : `uident` [ ( `ltac2_type` , ... , `ltac2_type` ) ]
          ltac2_fielddef         : [ mutable ] `ident` : `ltac2_type`

   Aliases are names for a given type expression and are transparently
   unfoldable to it. They cannot be recursive. The non-terminal
   :token:`qualid` designates identifiers starting with an uppercase.

   Variants are sum types defined by constructors and eliminated by
   pattern-matching. They can be recursive, but the `rec` flag must be
   explicitly set. Pattern matching must be exhaustive.

   .. todo PR next sentence: is that right?

>    Open variants can be extended with additional constructors using the `::=` form.

Reviewer 3: Apparently yes.

   Records are product types with named fields and eliminated by projection.

>    Likewise they can be recursive if the `rec` flag is set.

Reviewer 3: I guess mention of `with` for mutually recursive type definitions should be added.

.. cmd:: Ltac2 @ external @ident : @tac2type5 := @string @string
   :name: Ltac2 external

   This command defines an abstract type. It has no use for the end user and

>    is dedicated to types representing data coming from the OCaml world.

Reviewer 3: From what I can see in the code `Ltac2 @ external` is always used to define abstract Ltac2 terms (generally functions) and never Ltac2 types (which can already be declared abstract with `Ltac2 Type` without `:=`).

> Term Syntax

Reviewer 3: I'm surprised to see the term syntax being removed from this section and reintroduced under `Ltac2 Eval`. I think it should be kept here and it would be preferable to have the names of Ltac2 nonterminals be kept as close as possible to what they were before this PR.

Author: My mistake.  Most of the name changes are from changes in `g_ltac2.mlg`.  Also some local symbols need to be prefixed with `ltac2_` to make them unique.

.. cmd:: Ltac2 Type {? rec } @tac2typ_def {* with @tac2typ_def }
   :name: Ltac2 Type

   .. insertprodn tac2typ_def tac2rec_field

   .. prodn::
      tac2typ_def ::= {? @tac2typ_prm } @qualid {? {| := | ::= } @tac2typ_knd }
      tac2typ_prm ::= @ltac2_typevar
      | ( {+, @ltac2_typevar } )
      tac2typ_knd ::= @ltac2_type
      | [ {? {? %| } {+| @tac2alg_constructor } } ]
      | [ .. ]
      | %{ {? {+; @tac2rec_field } {? ; } } %}
      tac2alg_constructor ::= @ident
      | @ident ( {*, @ltac2_type } )
      tac2rec_field ::= {? mutable } @ident : @ltac2_type

   :n:`:=`
     Defines a type with a manifest

   :n:`::=`
     Extends an existing open variant type, a special kind of variant type whose constructors are not
     statically defined, but can instead be extended dynamically. A typical example
     is the standard `exn` type for exceptions. Pattern matching on open variants must always
     include a catch-all clause. They can be extended with this form, in which case
     :token:`tac2typ_knd` should be in the form :n:`[ {? {? %| } {+| @tac2alg_constructor } } ]`.

     .. todo PR above: does it add an additional constructor for an existing o v type or

>         does it create both the type (and apparently the constructor?

Reviewer: `Ltac2 Type foo ::= [ Foo ]` gives `Error: Unbound type foo`.  So it only creates the constructor, the type must already exist, as in https://github.com/coq/coq/blob/e6d92a9765f84c80f8c6a102fe5480490c747313/user-contrib/Ltac2/Init.v#L38

Reviewer: This seems not yet resolved?  I'm going to unresolve it, but feel free to drop a comment about why it's resolved and re-resolve it.

Author: I think the words a few lines above cover it: "Extends an **existing** open variant type"

   Without :n:`{| := | ::= }`
     Defines an abstract type for use representing data from OCaml.  Not for
     end users.

   :n:`with @tac2typ_def`
     Permits definition of mutually recursive type definitions.

   Each production of :token:`tac2typ_knd` defines one of four possible kinds

>    of definitions, respectively: alias, variant, open variant and record types.

Reviewer: ```suggestion
   of definitions, respectively: alias, variant, open variant, and record types.
```

Author: We're not using the Oxford comma in the doc.

   Aliases are names for a given type expression and are transparently

>    unfoldable to it. They cannot be recursive.

Reviewer: ```suggestion
   unfoldable to that expression. They cannot be recursive.
```

Reviewer: Resolved because you disagree with the suggestion?

Author: Resolved because I changed the document in the "More changes" commit.  I think the "Outdated" tells you that this part of the document has been changed:

![image](https://user-images.githubusercontent.com/1253341/89828748-53318200-db0e-11ea-98df-a4fc0468dd02.png)

   .. The non-terminal :token:`uident` designates identifiers starting with an uppercase.

   Variants are sum types defined by constructors and eliminated by
   pattern-matching. They can be recursive, but the `rec` flag must be
   explicitly set. Pattern matching must be exhaustive.

   Open variants can be extended with additional constructors using the `::=` form.

   Records are product types with named fields and eliminated by projection.
   Likewise they can be recursive if the `rec` flag is set.

.. cmd:: Ltac2 @ external @ident : @ltac2_type := @string @string
   :name: Ltac2 external

   Declares abstract terms.  Frequently, these declare OCaml functions
   defined in |Coq| and give their type information.  They can also declare

>    data structures from OCaml.  It has no use for the end user.

Reviewer: Minor wording suggestion (to correct the tense misagreement between sentences and add a bit of clarity):
```suggestion
   data structures from OCaml.  This construction has no use for the end user.
```

APIs
~~~~

Ltac2 provides over 150 API functions that provide various capabilities.  These
are declared with :cmd:`Ltac2 external` in :n:`lib/coq/user-contrib/Ltac2/*.v`.

> For example, `Message.print` defined in `Message.v` is used to print messages:

Reviewer: Why not
```suggestion
For example, :g:`Message.print` defined in `Message.v` is used to print messages:
```
to be a bit more semantically correct? (`g` for Gallina-like expressions, right?)

Also, is there any way to link `Message.v` to the stdlib doc for that file?  (And maybe do the same with `lib/coq/user-contrib/Ltac2/*.v` above)  cc @Zimmi48 @cpitclaudel ?

Reviewer 3: Unfortunately not yet. And I'm not sure we're gonna keep `:g:` in the future given that the semantic information is not used. So I don't personally use it.

.. coqtop:: none

   Goal True.

.. coqtop:: all abort

   Message.print (Message.of_string "fully qualified calls").
   From Ltac2 Require Import Message.
   print (of_string "unqualified calls").

Term Syntax
~~~~~~~~~~~

The syntax of the functional fragment is very close to that of Ltac1, except
that it adds a true pattern-matching feature, as well as a few standard
constructs from ML.

> In practice, there is some additional syntactic sugar that allows e.g. to

Reviewer: ```suggestion
In practice, there is some additional syntactic sugar that allows the user to e.g.
```

Reviewer: I think this correction is needed for the sentence to have correct grammar, so I'm unresolving it, as the change doesn't seem to have been integrated

Author: Already fixed in the "More changes commit".

Ltac Definitions
~~~~~~~~~~~~~~~~

.. cmd:: Ltac2 {? mutable } {? rec } @tac2def_body {* with @tac2def_body }
   :name: Ltac2

   .. insertprodn tac2def_body ltac2_input_fun

   .. prodn::
      tac2def_body ::= {| _ | @ident } {* @ltac2_input_fun } := @tac2expr6

>       ltac2_input_fun ::= @tac2pat0

Reviewer 3: This indirection looks awkward to me.

>    This command defines a new global Ltac2 value.

Reviewer 3: Since it is now documented in the syntax, the usual possibility to use syntactic sugar to put arguments of a function before the `:=` (and thus to omit the `fun` keyword) should be mentioned in the text.

In general, quotations can be introduced in terms using the following syntax, where

> :production:`quotentry` is some parsing entry.

Reviewer 3: This sentence does not make sense without the `prodn` that was deleted below. It should be adjusted.

.. _ltac2_built-in-quotations:

Built-in quotations
+++++++++++++++++++

.. insertprodn ltac2_quotations ltac1_expr_in_env

.. prodn::
   ltac2_quotations ::= ident : ( @lident )
   | constr : ( @term )
   | open_constr : ( @term )
   | pattern : ( @cpattern )

>    | reference : ( {| & @ident | @qualid } )

Reviewer 3: Is it feasible to remove this whitespace since the text says it's not permitted?
```suggestion
   | reference : ( {| &@ident | @qualid } )
```

Author: The grammar allows whitespace here, so I removed the text.

```
  globref:
    [ [ "&"; id = Prim.ident -> { CAst.make ~loc (QHypothesis id) }
      | qid = Prim.qualid -> { CAst.make ~loc @@ QReference qid }
    ] ]
  ;
```

Reviewer 3: @ppedrot Can you comment on why this text was here? It's possible that the grammar is lying here.

Reviewer: ping @ppedrot can you comment on the space in the grammar here? (submitting this as a comment review so that it shows up at the bottom of the PR and is not as easily missed)

Author: I'm not following the question anymore.

@Zimmi48 You're asking about the presence of "which parses either as :n:`@qualid` or :n:`&@ident`.", since removed??

> It's possible that the grammar is lying here.

Your evidence?

@JasonGross You're referring to the spaces here?  You think it should be just "ident:(" without spaces?  It's correct as it is.  Since the grammar is inserted automatically, eliminating the space after "ident" requires a a code change.  Not worth doing here.  Removing the space in ": (" is easier, but it doesn't seem worth it to me.

![image](https://user-images.githubusercontent.com/1253341/89838907-4ec29480-db21-11ea-8cd7-85ffb5160454.png)

   | ltac1 : ( @ltac1_expr_in_env )
   | ltac1val : ( @ltac1_expr_in_env )
   ltac1_expr_in_env ::= @ltac_expr
   | {* @ident } |- @ltac_expr

> The current implementation recognizes the following built-in quotations:

Reviewer 3: In the list below, the `ltac1` / `ltac1val` quotations are missing. They should be added with a link to where they are documented (subsections "Simple API" and "Low-level API" of "Ltac1 from Ltac2" section).

Syntax
++++++

> One can also insert Ltac2 code into Coq terms, similarl to what is possible in

Reviewer 3: Typo introduced

:g:`ltac2:(Control.refine (fun () => x))`, up to re-typechecking. It allows to
insert in a concise way an Ltac2 variable of type :n:`constr` into a Coq term.

Match
~~~~~

.. tacn:: match @tac2expr5 with {? @ltac2_branches } end
   :name: match (Ltac2)

   .. insertprodn ltac2_match atomic_tac2pat

   .. prodn::

>       ltac2_match ::= match @tac2expr5 with {? @ltac2_branches } end

Reviewer 3: If this `match` is mostly an implementation-detail, then it should probably not be presented first, without explanations. Also, the repetition between the `tacn` and the first line of the `prodn` is puzzling. The former should probably be removed.

Author: I think it's roughly the same as an OCaml match.  I'll move it down.

> Also, the repetition between the `tacn` and the first line of the `prodn` is puzzling. The former should probably be removed.

That makes it unfindable in the tactic index.

Match over terms
~~~~~~~~~~~~~~~~

Ltac2 features a construction similar to Ltac1 :tacn:`match` over terms, although
in a less hard-wired way.

.. tacn:: @ltac2_match_key @ltac2_expr__term with @ltac2_match_list end
   :name: lazy_match!; match!; multi_match!

   .. insertprodn ltac2_match_key ltac2_match_pattern

   .. prodn::
      ltac2_match_key ::= lazy_match!
      | match!
      | multi_match!
      ltac2_match_list ::= {? %| } {+| @ltac2_match_rule }
      ltac2_match_rule ::= @ltac2_match_pattern => @ltac2_expr
      ltac2_match_pattern ::= @cpattern
      | context {? @ident } [ @cpattern ]

   Evaluates :n:`@ltac2_expr__term`, which must yield a term, and matches it
   sequentially with the :token:`ltac2_match_pattern`\s, which may have
   metavariables.  When a match is found, metavariable values are substituted
   into :n:`@ltac2_expr`, which is then applied.

   Matching may continue depending on whether  `lazy_match!`, `match!` or `multi_match!`
   is specified.

   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but

>       I don't when which is required

Reviewer: Examples?

Author: Consider these examples in the text.  In the first one, 'h' is matched and its value is used.  It doesn't work with `?h`.  The second one only works with `?t` but not `t`.

```
      Goal forall A B : Prop, A -> B -> (A->B).
      intros.
      match! goal with
      | [ h : _ |- _ ] => let h := Control.hyp h in print (of_constr h); apply $h
      end.

      Ltac2 f2 t := match! t with
                    | context [ (~ ?t) ] => print (of_constr t); fail
                    | _ => ()
                    end.
```

Author: And the '$h' form is needed to access the value in tactic expressions but not otherwise.

Reviewer: The rule is that hypothesis names are bound without needing to be preceded by a `?`.  (Even in Ltac1, locations that bind only names never need a `?`, for example `let x := _ in _` binds `x`, while `let _ := x in _` does not.  Ltac2 does not support binding names like this in terms, so the only place where this rule shows up is when matching hypothesis names.)  Unification variables in terms/types require `?` to distinguish the unification variable from potential globals of the same name.

> And the '$h' form is needed to access the value in tactic expressions but not otherwise.

That's correct.  The `$` is used for antiquotations, and this is not specific to variables bound via matching.  That is, the need to use `$` is orthogonal from whether the variable was bound via `?` or whether it was bound via an Ltac2 `let`.

Reviewer 2: Antiquotations (the `$h` form) only exist in some notation scopes, like e.g. `constr`, `ident`, and more complex ones built out of those. Since `apply` is a notation that parses a `constr` as an argument, writing `h` there means the term variable `h` while `$h` is the Ltac2 bound variable `h` from the context.

Author: > The rule is that hypothesis names are bound without needing to be preceded by a `?`.

Yeah, and I recall now that Ltac1 does that, too.  OTOH, wouldn't it be more consistent to always require `?`, even for hypothesis names?

> @ppderot Antiquotations (the $h form) only exist in some notation scopes, like e.g. constr, ident, and more complex ones built out of those.

Is that the full list of notation scopes suitable for dropping into the chapter somewhere?

Reviewer: > Yeah, and I recall now that Ltac1 does that, too. OTOH, wouldn't it be more consistent to always require `?`, even for hypothesis names?

Maybe... there are a couple of issues with this, though.  The first issue is that this would suggest that something like `?h : ?t, ?p : ?h = ?h |- _` would bind `t` to the type of `h`, but this is not the case, as there is check that terms bound in patterns match bound hypotheses.  The second issue is that this would require changing a great deal of code to add an extra character in many places (at least if this change were required in Ltac1 as well).  I'm not convinced, but feel free to open an issue about it.

Author: I'll let it be for now.  Ltac1 has other things that would be a higher priority to improve.

   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas
   in the :n:`@ltac2_expr`\s, the question mark is omitted.  Choose your metavariable
   names with care to avoid name conflicts.  For example, if you use the metavariable `S`,
   then the :token:`ltac2_expr` can't use `S` to refer to the constructor of `nat`

>    without qualifying the constructor as `Datatypes.S`.

Reviewer: Ltac2 variable names are not allowed to be capital letters.  So if your metavariable starts with a capital letter, you cannot use it at all.  Additionally, unlike in Ltac1, I believe conflicts are *not* an issue, because you have to antiquote Ltac2 variables.  So if you name your metavariable `?nat`, there is still a difference between `constr:(nat)` and `constr:($nat)` (or whatever the correct anti-quotation is; I can't quite recall offhand).

But this is already covered below with
```
Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
bindings, so that there will be a syntax error if one of the bound variables
starts with an uppercase character.
```

Author: I dropped the 3 sentences starting with "Choose ...".  The first sentence is the same question as above about when a '?' is required on a variable name.

   .. todo below: is matching non-linear unification?  is it the same or different

>       from unification elsewhere in Coq?

Reviewer: This is a question for @ppedrot ; I have not tested the non-linearity of Ltac2 against Ltac1

Author: I raised the question for Ltac1 but didn't get a useful answer.

Reviewer 2: It's the same underlying algorithm as Ltac1, and IIRC I've explained it on another issue, but for the sake of completeness:
- inside the same term, e.g. `match goal with [ |- ?P -> ?P ] => ... end` or `match goal with [ H : ?P -> ?P |- _ ] => ... end`  non-linear patterns are compared up to syntactic equality.
- across hypotheses, e.g. e.g. `match goal with [ H : ?P |- ?P ] => non-linear pattern are compared up to convertibility.

Author: @ppedrot Thanks.  I originally asked the question while working on the Ltac1 chapter.  Indeed, the text here is a cut-and-paste from that chapter.

> non-linear patterns are compared up to syntactic equality.

I don't understand "up to ...".  Do you mean it's a strictly syntactic comparison of the ASTs, i.e. "3" doesn't match "1+2" or is it more powerful than that?  And I expect the "non-linear" is not restrictive i.e. linear and non-linear patterns are processed the same way.  Right?

> non-linear pattern are compared up to convertibility.

Most unfortunately, I don't see a definition of "convertibility" anywhere in the doc.  It seems to be an essential concept, though, and if so, I will look the other way while I drop the term into the text, with the expectation we'll add a definition in the near future.

Your words mean: if (A is convertible into B or vice-versa **OR** A and B both convert to the same canonical form) then it's a match?

Reviewer 2: > it's a strictly syntactic comparison

Syntactic equality includes α-conversion though, so e.g. it'll match `fun x : nat => x` with `fun y : nat => y`

> I expect the "non-linear" is not restrictive

It doesn't make sense to ask for the comparison algorithm for a linear pattern, since it'll never be called. But I guess it's a matter of phrasing.

>  I don't see a definition of "convertibility" anywhere in the doc. It seems to be an essential concept

It definitely is, and even has [its own section](https://coq.inria.fr/refman/language/core/conversion.html) in the documentation of typing rules.

> Your words mean

Conversion is an equivalence relation on well-typed terms, so you don't need the OR, just that "A is convertible to B".

Author: I don't see a definition of α-conversion anywhere in the documentation.  Seems like that should be in the conversion chapter, right?  I suppose we include α-conversion for almost every comparison of terms.  Should α be included in phrases such as: "We say that two terms t1 and t2 are βδιζη-convertible, or simply convertible, or equivalent"?

> It doesn't make sense to ask for the comparison algorithm for a linear pattern, since it'll never be called. But I guess it's a matter of phrasing.

Yeah, basic writing: don't introduce unnecessary terms.  If I say "non-linear pattern" that immediately suggests the question of how linear patterns are handled.  Of course that's the trivial case.  Not worth an extra sentence to cover that detail and would be a bit harder to understand.

Not sure how I missed the convertibility section.  I did a full-text search of the source and didn't find it.  Pilot error or an IDE glitch.

> Conversion is an equivalence relation on well-typed terms, so you don't need the OR, just that "A is convertible to B".

Is that the most common phrase?  From the description (below) "A and B are convertible" sounds more accurate to me.

> We say that two terms t1 and t2 are βδιζη-convertible, or simply convertible, or equivalent, in the global environment E and local context Γ iff there exist terms u1 and u2 such that E[Γ]⊢t1▹…▹u1 and E[Γ]⊢t2▹…▹u2 and either u1 and u2 are identical up to irrelevant subterms, or they are convertible up to η-expansion, i.e. u1 is λx:T. u′1 and u2x is recursively convertible to u′1, or, symmetrically, u2 is λx:T. u′2 and u1x is recursively convertible to u′2. We then write E[Γ]⊢t1=βδιζηt2.

Author: Also, I happened to notice this in https://coq.inria.fr/distrib/V8.12+beta1/stdlib/Ltac2.Constr.html

```
Ltac2 @ external equal : constr -> constr -> bool := "ltac2" "constr_equal".
Strict syntactic equality: only up to α-conversion and evar expansion
```

Is evar expansion perhaps also used for for unification or not?  If so, does it correspond to one of `βδιζη`?

   matching is second-order and returns a functional term.

   .. todo the `@?ident` form is in dangling_pattern_extension_rule, not included in the doc yet
      maybe belongs with "Applications"

   `lazy_match!`
      Causes the match to commit to the first matching branch
      rather than trying a new match if :n:`@ltac2_expr` fails.
      :ref:`Example<ltac2_match_vs_lazymatch_ex>`.

   `match!`
      If :n:`@ltac2_expr` fails, continue matching with the next branch.
      Failures in subsequent tactics (after the `match!`) will not cause selection
      of a new branch.  Examples :ref:`here<ltac2_match_vs_lazymatch_ex>` and
      :ref:`here<ltac2_match_vs_multimatch_ex>`.

   `multi_match!`
      If :n:`@ltac2_expr` fails, continue matching with the next branch.
      When a :n:`@ltac2_expr` succeeds for a branch, subsequent failures
      (after the `multi_match!`) causing consumption of all the successes
      of :n:`@ltac2_expr` trigger selection of a new matching branch.
      :ref:`Example<ltac2_match_vs_multimatch_ex>`.

   :n:`@cpattern`
      The syntax of :token:`cpattern` is
      the same as that of :token:`term`\s, but it can contain pattern matching

>       metavariables in the form :n:`?@ident`.  :g:`_` can be used to match

Reviewer: It can also contain `@?foo`, no?

Author: Yes.  Can you explain what it means?  And if so, can you provide an example?  At @Zimmi48's suggestion, I omitted the relevant production from the grammar for `term10`.  (BTW there's still a good bit of syntax that we don't explain such as in the `term*` productions.  Nor do we fully explain patterns anywhere.  We'll address these in future PRs.)

Reviewer: I believe I already explained this in the Ltac1 PR, let met try to find it

Reviewer: Quoting https://github.com/coq/coq/pull/12103#discussion_r436297754:
> 
> As for the second paragraph, here's the same tactic but written using second-order pattern matching:
> ```coq
> Ltac uncurry f :=
>   lazymatch (eval cbv beta in f) with
>   | fun (a : ?A) (b : ?B) => @?f a b
>     => uncurry (fun ab : A * B => f (fst ab) (snd ab))
>   | ?f => f
>   end.
> ```
> 
> Perhaps a better example of the former is swapping args:
> [...]
> But you can also do with with a neat trick of `@?` patterns:
> ```coq
> Ltac swap_args f :=
>   lazymatch f with
>   | fun a b => @?f b a (* note the reversed order *)
>     => f
>   end.
> 
> Goal True.
>   let f := swap_args (fun x y => x + y) in
>   pose f. (* fun y x : nat => x + y : nat -> nat -> nat *)
> ```

I'll let you convert these to Ltac2, let me know if you have any trouble with the conversion.

Author: Now I recall the Ltac1 discussion.  I will leave this for a future PR.

Reviewer: Maybe just add 
```suggestion
      metavariables in the form :n:`?@ident` and :n:`@?@ident`.  :g:`_` can be used to match
```
?

Reviewer: And maybe add a todo referencing the comment above?

      irrelevant terms.

      .. todo Example is broken :ref:`Example<ltac2_match_with_holes_ex>`.

      .. todo Didn't understand the following 2 paragraphs well enough to revise
         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
         possible example

      When a metavariable in the form :n:`?id` occurs under binders,
      say :n:`x__1, …, x__n` and the expression matches, the
      metavariable is instantiated by a term which can then be used in any
      context which also binds the variables :n:`x__1, …, x__n` with
      same types. This provides with a primitive form of matching under

>       context which does not require manipulating a functional term.

Reviewer: I thought Ltac2 failed to handle open terms entirely?  cc @ppedrot

Author: This is a cut-and-paste from Ltac1.  Perhaps it should be removed here?

Reviewer: Yes.  Or just say something like "Unlike Ltac1, Ltac2 :n:`?id` metavariables only match closed terms." and drop a link to the corresponding paragraph in the Ltac1 documentation?

Author: I added that wording.  Ltac1 has the identical paragraph, so not sure what to link to.  Furthermore, we don't even define open and closed terms.  (Open means there's a metavariable in it??)

      dependencies in the variables :n:`@ident__i` and returns a functional term
      of the form :n:`fun @ident__1 … @ident__n => @term`.

   .. _match_term_context:

   :n:`context {? @ident } [ @cpattern ]`
      Matches any term with a subterm matching :token:`cpattern`. If there is a match
      and :n:`@ident` is present, it is assigned the "matched
      context", i.e. the initial term where the matched subterm is replaced by a
      hole.  Note that `context`
      (with very similar syntax) appearing after the `=>` is the :tacn:`context` tactic.

      For :tacn:`match!` and :tacn:`multi_match!`, if the evaluation of the :token:`ltac2_expr`
      fails, the next matching subterm is tried. If no further subterm matches, the next branch
      is tried.  Matching subterms are considered from top to bottom and from left to
      right (with respect to the raw printing obtained by setting the
      :flag:`Printing All` flag).  :ref:`Example<ltac2_match_term_context_ex>`.

   .. todo There's a more realistic example from @JasonGross here:
      https://github.com/coq/coq/pull/12103#discussion_r432996954

   :n:`@ltac2_expr`

>       The tactic to apply if the construct matches.  Metavariable values from the pattern
>       match are substituted
>       into :n:`@ltac2_expr` before it's applied.  Note that metavariables are not
>       prefixed with the question mark as they are in :token:`cpattern`.

Reviewer: ```suggestion
      The tactic to apply if the construct matches.  Metavariable values from the pattern
      match are bound as Ltac2 variables in the context of :n:`@ltac2_expr` before it's
      applied.  Note that the names of bound metavariables are not prefixed with question
      marks as they are in :token:`cpattern`.
```
But why is this text here in the first place?  It's redundant with the below
```
Variables from the :n:`@tac2pat1` are statically bound in the body of the branch, to
values of type `constr` for the variables from the :n:`@term` pattern and to a
value of type `Pattern.context` for the variable :n:`@lident`.
```
and less accurate to boot.

Author: Check what I did with your suggested change.

I can't make sense of "values of type `constr` for the variables from the :n:`@term` pattern and to a value of type `Pattern.context` for the variable :n:`@lident`."  lident is in the pattern, right?

Reviewer: I believe it refers to, e.g., in `match! '(True /\ True) with | context c [and ?x] => ... end`, we'll get `c` bound with type `Pattern.context` and `x` bound with type `constr`.  Does this explain it?

Author: That helps.  I think this covers it:

Variables from the :n:`@tac2pat1` are statically bound in the body of the branch. Variables from the :n:`@term` pattern have values of type `constr`. Variables from the :n:`@ident` in the `context` construct have values of type `Pattern.context` (defined in `Pattern.v`).

      If :token:`ltac2_expr` evaluates to a tactic, then it is
      applied. If the tactic succeeds, the result of the match expression is
      :n:`()`.  If :token:`ltac2_expr` does not evaluate

>       to a tactic, that value is the result of the match expression.

Reviewer: What?  This paragraph makes no sense.  Was it taken directly from the Ltac1 chapter?  The semantics of Ltac1 and Ltac2 are quite different, and Ltac2 (thankfully) does not have two separate phases like Ltac1 does.

Author: > Was it taken directly from the Ltac1 chapter?

Pretty much.  As I said [here](url), "I rewrote the sections for `match!` and `match! goal` using the Ltac1 text as a starting point."  It's a real struggle guessing what Ltac2 is intended to do or how it's supposed to differ from Ltac1.  I hope you can tell me which aspects don't apply to Ltac2.

Reviewer: From a bird's eye view: I believe it took a while to convey how Ltac1 has two separate evaluation phases, and how it's used to both run tactics and return values through a weird semantics.  This is confusing, AFAICT, to literally everyone who encounters it.  Ltac2 removes all of this in favor of the OCaml way of doing things; mutating the goal is considered a side-effect, and otherwise the semantics are effectively the same as OCaml's (e.g., side-effect-only functions return unit, functions can have both a return value and side effects, etc).  The one complication is that effects on the goal live inside the backtracking monad, while other side-effects do not, as seen by:
```coq
Require Import Ltac2.Ltac2.
Require Ltac2.Control Ltac2.Int Ltac2.Message.

Ltac2 foo () :=
  let v := { contents := 0 } in
  let incr () := ((v.(contents) := Int.add (v.(contents)) 1);
                 Message.print (Message.of_int (v.(contents)))) in
  Control.plus incr (fun _ => incr ()); Control.zero Not_found.

Goal True.
  try (foo ()). (* 1 2 *)
```

Ltac2 is basically about exposing both Ltac1 features and the lower-level OCaml tactic API, but doing away with the home-brewed strange semantics of Ltac1.

Author: Yeah.  This chapter will need a couple more rounds of revision before it's really good.  My main goal in this round is to get the syntax right.

I have pondered whether we could evolve Ltac1 semantics so they're not so strange.  For example, could we:
- add the `$var` construct to Ltac1 (optional, not required)
- add an option that, at run time, reports constructs that lack the appropriate `$var`
- add an option to require use of `$var` and drop the related strange semantic
and change the defaults for the options over a release or two?

It seems to me there's a lot of Ltac1 code out there, that converting is non-trivial and it's not clear (at least to me) when Ltac2 will be really ready for prime time and well-documented.  Incremental improvements to Ltac1 along these lines seems like it would be less disruptive for users.

      If :n:`@ltac2_expr` is a tactic with backtracking points, then subsequent
      failures after a :tacn:`lazy_match!` or :tacn:`multi_match!` (but not :tacn:`match!`) can cause
      backtracking into :n:`@ltac2_expr` to select its next success.

   .. note::
      Each Ltac2 construct is processed in two phases: an evaluation phase and an execution phase.
      In most cases, tactics that may change the proof state are applied in the second phase.
      (Tactics that generate integer, string or syntactic values, such as :tacn:`fresh`,

>       are processed during the evaluation phase.)

Reviewer: What?  This is false, and does not apply to Ltac2, only to Ltac1.

      Unlike other tactics, `*match*!` tactics get their first success (applying tactics to do
      so) as part of the evaluation phase.  Among other things, this can affect how early

>       failures are processed in :tacn:`assert_fails`.  Please see the note in :tacn:`assert_fails`.

Reviewer: Again, this only applies to Ltac1.

Variables from the :n:`@tac2pat1` are statically bound in the body of the branch, to
values of type `constr` for the variables from the :n:`@term` pattern and to a
value of type `Pattern.context` for the variable :n:`@lident`.

Note that unlike Ltac, only lowercase identifiers are valid as Ltac2
bindings, so that there will be a syntax error if one of the bound variables
starts with an uppercase character.

The semantics of this construction are otherwise the same as the corresponding
one from Ltac1, except that it requires the goal to be focused.

.. _ltac2_match_vs_lazymatch_ex:

.. example:: Ltac2 Comparison of lazy_match! and match!

   (Equivalent to this :ref:`Ltac1 example<match_vs_lazymatch_ex>`.)

   These lines define a `msg` tactic that's used in several examples as a more-succinct
   alternative to `print (to_string "...")`:

>    .. From Ltac2 Require Import Message. *outside* the proof lest it disappear with "abort"

Reviewer: ?

Author: I was having trouble with `.. coqtop:: abort`.  Would be hard to reproduce at this point.

   .. coqtop:: in

      From Ltac2 Require Import Message.
      Ltac2 msg x := print (of_string x).

   .. coqtop:: none

      Goal True.

   In :tacn:`lazy_match!`, if :token:`ltac2_expr` fails, the :tacn:`lazy_match!` fails;
   it doesn't look for further matches.  In :tacn:`match!`, if :token:`ltac2_expr` fails
   in a matching branch, it will try to match on subsequent branches.

   .. coqtop:: all

>       Fail lazy_match! constr:(True) with

Reviewer: Do you want to mention somewhere that `'True` is a shorter form of `constr:(True)`?

Author: I added a few words.  This should be explained with the syntax for `ltac2_expr` is explained--currently none of that syntax is explained and won't be added in this PR.

      | True => msg "branch 1"; fail
      | _ => msg "branch 2"
      end.

      match! constr:(True) with
      | True => msg "branch 1"; fail
      | _ => msg "branch 2"
      end.

.. _ltac2_match_vs_multimatch_ex:

.. example:: Ltac2 Comparison of match! and multi_match!

   (Equivalent to this :ref:`Ltac1 example<match_vs_multimatch_ex>`.)

   :tacn:`match!` tactics are only evaluated once, whereas :tacn:`multi_match!`
   tactics may be evaluated more than once if the following constructs trigger backtracking:

   .. coqtop:: all

>       Fail match! constr:(True)with

Reviewer: ```suggestion
      Fail match! constr:(True) with
```

      | True => msg "branch 1"
      | _ => msg "branch 2"
      end ;
      msg "branch A"; fail.

   .. coqtop:: all

      Fail multi_match! constr:(True) with
      | True => msg "branch 1"
      | _ => msg "branch 2"
      end ;
      msg "branch A"; fail.

.. _ltac2_match_with_holes_ex:

> .. todo EXAMPLE DOESN'T WORK: Pattern matching?   Matching a pattern with holes

Reviewer: As I said above, I believe open terms do not work at all in Ltac2 yet

Author: We'll find out.  I hope you understand the example is commented out so it won't appear in the generated HTML.  It's a reminder for the future when open terms do work in Ltac2.

Reviewer: ```suggestion
.. todo EXAMPLE DOESN'T WORK: Ltac2 does not (yet?) handle pattern variables matching open terms.  Matching a pattern with holes
```

   (Equivalent to this :ref:`Ltac1 example<match_with_holes_ex>`.)

   Notice the :tacn:`idtac` prints ``(z + 1)`` while the :tacn:`pose` substitutes
   ``(x + 1)``.

   .. coqtop:: all

      match! constr:(fun x => (x + 1) * 3) with

>       | fun z => ?y * 3 => print (of_constr y); pose (fun z: nat => y * 5)

Reviewer: ```suggestion
      | fun z => ?y * 3 => print (of_constr y); pose (fun z: nat => $y * 5)
```
I think (but it's still not going to work, because (a) open terms don't work in Ltac2, and (b) names of functions are not bound in Ltac2, so the `z`'s don't line up anyway

Author: yeah, still doesn't work

   :name: lazy_match! goal; match! goal; multi_match! goal

   .. insertprodn goal_match_list gmatch_hyp_pattern

   .. prodn::
      goal_match_list ::= {? %| } {+| @gmatch_rule }
      gmatch_rule ::= @gmatch_pattern => @ltac2_expr
      gmatch_pattern ::= [ {*, @gmatch_hyp_pattern } |- @ltac2_match_pattern ]
      gmatch_hyp_pattern ::= @name : @ltac2_match_pattern

   Matches over goals, similar to Ltac1 :tacn:`match goal`.
   Use this form to match hypotheses and/or goals in the proof context.  These patterns have zero or
   more subpatterns to match hypotheses followed by a subpattern to match the conclusion.  Except for the
   differences noted below, this works the same as the corresponding :n:`@ltac2_match_key @ltac2_expr` construct
   (see :tacn:`match!`).  Each current goal is processed independently.

   :n:`{*, @gmatch_pattern }`
      Patterns to match with hypotheses.  Each pattern must match a distinct hypothesis in order
      for the branch to match.

      Hypotheses have the form :n:`@name {? := @term__binder } : @type`.  Currently Ltac2 doesn't
      allow matching on or capturing the value of :n:`@term__binder`.  It only supports matching on

>       the :token:`name` and the :token:`type`, for example `n : ?t`.

Reviewer: This seems like a rather severe restriction.  Is there an issue open for it?

Author: I didn't file one or check whether there already is one.  Feel free to file one.

Reviewer: Reported as https://github.com/coq/coq/issues/12803

      .. currently only supports the first row
         :list-table::
         :widths: 2 1
         :header-rows: 1

         * - Pattern syntax
           - Example pattern

         * - :n:`@name : @ltac2_match_pattern`
           - `n : ?t`

         * - :n:`@name := @match_pattern__binder`
           - `n := ?b`

         * - :n:`@name := @term__binder : @type`
           - `n := ?b : ?t`

>          * - :n:`@name := [ @match_pattern__binder ] : @ltac2_match_pattern`

Reviewer: This isn't part of the syntax you note above?  OR am I missing something?

Author: Nope.  Compare the statement to the syntax in the grammar.

Reviewer: Ah, I believe this is in your comment "currently only supports the first row".  Perhaps we should comment out the other rows?

Author: It is commented out.  Rst syntax is like Python, it goes by indentation.  Comments begin with `.. ` and end just before the first following line that is indented the same amount or less.

You may want to look at the HTML output, much easier to read.

           - `n := [ ?b ] : ?t`

         :token:`name` can't have a `?`.  Note that the last two forms are equivalent except that:

         - if the `:` in the third form has been bound to something else in a notation, you must use the fourth form.
           Note that cmd:`Require Import` `ssreflect` loads a notation that does this.
         - a :n:`@term__binder` such as `[ ?l ]` (e.g., denoting a singleton list after
           :cmd:`Import` `ListNotations`) must be parenthesized or, for the fourth form,
           use double brackets: `[ [ ?l ] ]`.

      If there are multiple :token:`gmatch_hyp_pattern`\s in a branch, there may be multiple ways to match them to hypotheses.
      For :tacn:`match! goal` and :tacn:`multi_match! goal`, if the evaluation of the :token:`ltac2_expr` fails,
      matching will continue with the next hypothesis combination.  When those are exhausted,
      the next alternative from any `context` constructs in the :token:`ltac2_match_pattern`\s is tried and then,

>       when the context alternatives are exhausted, the next branch is tried.

Reviewer: This confuses me a bit.  What if there are `context`s in the hypothesis types?  Will it try other combinations of hypotheses before trying the first combination with a different set of `context` matches?  That seems like very, very strange semantics.  (If this is only talking about the `context` constructs in goal, that is not clear in this text.)

Author: This is basically the same description as for Ltac1.  Submit an issue if you like.  Here's an Ltac1 example:

```
Goal forall A B : Prop, A -> ((~ True) <> (~ False)) -> (A->B).
intros A B H.
match goal with
| H1 : _, H2 : _ |- context [ (~ ?t) ] => idtac "match " H1 H2 t; fail
| _ => idtac
end.

match  B H ((~ True) = (~ False))
match  A H ((~ True) = (~ False))
match  H B ((~ True) = (~ False))
match  A B ((~ True) = (~ False))
match  H A ((~ True) = (~ False))
match  B A ((~ True) = (~ False))
match  B H True
match  A H True
match  H B True
match  A B True
match  H A True
match  B A True
match  B H False
match  A H False
match  H B False
match  A B False
match  H A False
match  B A False
```

      :ref:`Example<ltac2_match_goal_multiple_hyps_ex>`.

   `reverse`
      Hypothesis matching for :token:`gmatch_hyp_pattern`\s normally begins by matching them from left to right,
      to hypotheses, last to first.  Specifying `reverse` begins matching in the reverse order, from
      first to last.  :ref:`Normal<ltac2_match_goal_hyps_ex>` and :ref:`reverse<ltac2_match_goal_hyps_rev_ex>` examples.

   :n:`|- @ltac2_match_pattern`
      A pattern to match with the current goal

>    Note that unlike Ltac, only lowercase identifiers are valid as Ltac2

Reviewer: ```suggestion
   Note that unlike Ltac1, only lowercase identifiers are valid as Ltac2
```

Note also that this is duplicated text; is that intentional?

Author: Yes, intentional.  One is in match over terms, the other is in match over goals.  Readers should not have to read both to glean this fact, so some repetition is beneficial.  A long repetition should be replaced with shared and hyperlinked text.

>    bindings, so that there will be a syntax error if one of the bound variables

Reviewer: Is this actually true, or does the bound variable just become unusable?

Author: An error, but not a syntax error:

```
   Ltac2 msg S := print (of_string S).

Unbound constructor S
Raised at file "clib/exninfo.ml", line 80, characters 4-11
Called from file "user-contrib/Ltac2/tac2intern.ml", line 569, characters 11-34
```

         fail
      | [ |- _ ] => ()
      end.

Match on values
~~~~~~~~~~~~~~~

.. tacn:: match @ltac2_expr5 with {? @ltac2_branches } end
   :name: match (Ltac2)

   Matches a value, akin to the OCaml `match` construct.  By itself, it doesn't cause backtracking
   as do the `*match*!` and `*match*! goal` constructs.

   .. insertprodn ltac2_branches atomic_tac2pat

   .. prodn::
      ltac2_branches ::= {? %| } {+| @tac2pat1 => @ltac2_expr }
      tac2pat1 ::= @qualid {+ @tac2pat0 }
      | @qualid
      | [ ]
      | @tac2pat0 :: @tac2pat0
      | @tac2pat0
      tac2pat0 ::= _
      | ()
      | @qualid
      | ( {? @atomic_tac2pat } )
      atomic_tac2pat ::= @tac2pat1 : @ltac2_type
      | @tac2pat1 , {*, @tac2pat1 }
      | @tac2pat1

.. note::

   For now, deep pattern matching is not implemented.

.. _ltac2_notations:

Notations
---------

.. cmd:: Ltac2 Notation @string {* @name ( @ltac2_scope ) } {? : @int } := @ltac2_expr
   :name: Ltac2 Notation

   .. todo seems like name maybe should use lident rather than ident, considering:

      Ltac2 Notation "ex1" X(constr) := print (of_constr X).
      ex1 1.

      Unbound constructor X

>       This works fine with lower-case "x" in place of "X"

Reviewer: Indeed, maybe submit an issue for this?

Author: The user-friendly thing would be to give a user-friendly error message like "Ltac2 identifiers must begin with a lower-case letter".  https://github.com/coq/coq/issues/12804

   .. todo Ltac2 Notation := permits redefining same symbol (no warning)
      Also allows defining a symbol beginning with uppercase, which is prohibited
      in similar constructs.

   :cmd:`Ltac2 Notation` provides a way to extend |Coq|'s syntax.  The left-hand
   side (before the `:=`) defines the syntax to recognize and gives formal parameter
   names for the syntactic values.  When the notation is used, the values are substituted
   into the right-hand side.  In the following example, `x` is the formal parameter name, `constr` is
   the name of a nonterminal describing the syntax (which the documentation calls :token:`term`).
   `print` and `of_constr` are functions provided by |Coq| through `Message.v`.

>    .. todo "print" doesn't seem to pay attention to "Set Printing All"

Reviewer: I don't think it's `print` at fault, but `Message.of_constr`, no?

Author: That may be, but I think the comment is fine as it is.  I can only read so much code while updating the doc.

   .. example:: Printing a :n:`@term`

      .. coqtop:: none

         Goal True.

      .. coqtop:: all

         From Ltac2 Require Import Message.
         Ltac2 Notation "ex1" x(constr) := print (of_constr x).
         ex1 (1 + 2).

   There are also metasyntactic classes described :ref:`here<syntactic_classes>`
   that combine other items.  For example, `list1(constr, ",")`
   recognizes a comma-separated list of one or more :token:`term`\s.  (The `auto` on the right-hand

>    side is contrived; currently there's no way to print a list, nor, by the way, are there conversions

Reviewer: What?  You can iterate over the list by hand and build the message to print it yourself.  Admittedly, this is kind-of annoying, and is a missing feature.

Author: I don't know there was a way to do (`Message.concat`) until recently.

   :name: Ltac2 Notation

   .. todo JEF explain components, e.g. string and int

   .. todo seems like syn_node maybe should use lident rather than ident, considering:

      Ltac2 Notation "ex1" X(constr) := print (of_constr X).
      ex1 1.

      Unbound constructor X

      This works find with lower-case "x" in place of "X"

   .. todo Ltac2 Notation := permits redefining same symbol (no warning)
      Also allows defining a symbol beginning with uppercase, which is prohibited
      in similar constructs.

   :cmd:`Ltac2 Notation` provides a way to extend |Coq|'s syntax.  The left-hand
   side (before the `:=`) defines the syntax to recognize and gives formal parameter
   names for the syntactic values.  When the notation is used, the values are substituted
   into the right-hand side.  In the following example, `x` is the formal parameter name, `constr` is
   the name of a nonterminal describing the syntax (which the documentation calls :token:`term`).
   `print` and `of_constr` are functions provided by |Coq| through `Message.v`.

   .. todo "print" doesn't seem to pay attention to "Set Printing All"

>    EXAMPLE FAILS IN COQTOP DIRECTIVE, WORKS IN coqtop/CoqIDE::

Reviewer 3: Any way these examples could be done without ever calling `reset` in all the Ltac2 chapter?

Author: Yeah, I figured that out the other day.  (I had not understood that `abort` was applied after the following block of Coq code.

   an effectful language such as Ltac2, any syntactic argument to an abbreviation
   is thunked on-the-fly during its expansion.

   For instance, suppose that we define the following.

   :n:`Ltac2 Notation foo := fun x => x ().`

   Then we have the following expansion at internalization time.

   :n:`foo 0 ↦ (fun x => x ()) (fun _ => 0)`

   Note that abbreviations are not type checked at all, and may result in typing
   errors after expansion.

.. _syntactic_classes:

Syntactic classes
~~~~~~~~~~~~~~~~~

The simplest syntactic classes in Ltac2 notations represent individual nonterminals
from the |Coq| grammar.  Only a few selected nonterminals are available as syntactic classes.
In addition, there are metasyntactic operations for describing
more complex syntax, such as making an item optional or representing a list of items.
When parsing, each syntactic class expression returns a value that's bound to a name in the
notation definition.

Metasyntactic classes are described with a form of S-expression:

   .. insertprodn ltac2_scope syn_node

   .. prodn::
      ltac2_scope ::= @string
      | @int
      | @syn_node
      | @syn_node ( {+, @ltac2_scope } )
      syn_node ::= _

>       | @ident

Reviewer 3: Any reason to define `syn_node` instead of reusing the common `name`?

Author: It's from the Ltac2 grammar.  I will change.

    Used for semantic effect only, parses the same as :token:`ltac2_scope`.
    If :n:`e` is the parsed expression for :token:`ltac2_scope`, `thunk`
    returns :n:`fun () => e`.

  :n:`@string`
    parses the corresponding string as an identifier and returns :n:`()`.

  .. todo PR I dropped "terminal" and "keyword" because they don't seem to behave
     any differently from just giving a string, e.g. for the keyword "match",
     x(terminal("match")) worked the same as x("match")

A few syntactic classes contain antiquotation features. For the sake of uniformity, all
antiquotations are introduced by the syntax :n:`$@lident`.

A few other specific syntactic classes exist to handle Ltac1-like syntax, but their use is
discouraged and they are thus not documented.

For now there is no way to declare new syntactic classes from the Ltac2 side, but this is
planned.

Other nonterminals that have syntactic classes are listed here.
The class name is the Ltac2 nonterminal name without the prefix :n:`q_`.  (Exception:
the class name for :token:`q_strategy_flag` is :n:`q_strategy`).

   .. list-table::
      :header-rows: 1

      * - Ltac2 nonterminal
        - Equivalent

>         - Notes

Reviewer 3: Do you intend to keep this third column?

Author: Probably not but won't remove it quite yet.

   .. prodn::
      ltac2_match_key ::= lazy_match!
      | match!
      | multi_match!
      ltac2_match_list ::= {? %| } {+| @ltac2_match_rule }
      ltac2_match_rule ::= @ltac2_match_pattern => @ltac2_expr
      ltac2_match_pattern ::= @cpattern
      | context {? @ident } [ @cpattern ]

   Evaluates :n:`@ltac2_expr__term`, which must yield a term, and matches it
   sequentially with the :token:`ltac2_match_pattern`\s, which may contain
   metavariables.  When a match is found, metavariable values are substituted
   into :n:`@ltac2_expr`, which is then applied.

   Matching may continue depending on whether  `lazy_match!`, `match!` or `multi_match!`
   is specified.

   .. todo PR I see ? is sometimes need in the ltac2_expr and sometimes not, but
      I don't when which is required

   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas
   in the :n:`@ltac2_expr`\s, the question mark is omitted.

   .. todo below: is matching non-linear unification?  is it the same or different
      from unification elsewhere in Coq?

   Matching is non-linear: if a
   metavariable occurs more than once, each occurrence must match the same

>    expression.  Matching is first-order except on variables of the form :n:`@?@ident`

Reviewer: Does `@` need to be escaped in `@?`?

Author: Nope.

![image](https://user-images.githubusercontent.com/1253341/89831649-0ac89300-db13-11ea-98a1-24c938cd1757.png)

BTW, you can download the HTML doc by clicking on "Details" near the bottom of the page.

![image](https://user-images.githubusercontent.com/1253341/89831758-38154100-db13-11ea-8dcf-d36ba4475bd3.png)

         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
         possible example

      When a metavariable in the form :n:`?id` occurs under binders,
      say :n:`x__1, …, x__n` and the expression matches, the
      metavariable is instantiated by a term which can then be used in any
      context which also binds the variables :n:`x__1, …, x__n` with
      same types. This provides with a primitive form of matching under
      context which does not require manipulating a functional term.

      There is also a special notation for second-order pattern matching: in an
      applicative pattern of the form :n:`@?@ident @ident__1 … @ident__n`,
      the variable :token:`ident` matches any complex expression with (possible)
      dependencies in the variables :n:`@ident__i` and returns a functional term
      of the form :n:`fun @ident__1 … @ident__n => @term`.

   .. _match_term_context:

   :n:`context {? @ident } [ @cpattern ]`
      Matches any term with a subterm matching :token:`cpattern`. If there is a match
      and :n:`@ident` is present, it is assigned the "matched
      context", i.e. the initial term where the matched subterm is replaced by a

>       hole.  Note that `context`
>       (with very similar syntax) appearing after the `=>` is the :tacn:`context` tactic.

Reviewer: There is no `context` tactic in Ltac2.
```suggestion
      hole.  This matched context has type `Pattern.context` and the hole
      can be filled with the function `Pattern.instantiate`.
```

Author: Your quotes are fine.  I said:

> This hole in the matched context can be filled with the expression
      :n:`Pattern.instantiate @ident @cpattern`.

.. _ltac2_match_goal_hyps_ex:

.. example:: Ltac2 Matching hypotheses

   (Equivalent to this :ref:`Ltac1 example<match_goal_hyps_ex>`.)

   Hypotheses are matched from the last hypothesis (which is by default the newest
   hypothesis) to the first until the :tacn:`apply` succeeds.

   .. coqtop:: all abort

      Goal forall A B : Prop, A -> B -> (A->B).
      intros.
      match! goal with
      | [ h : _ |- _ ] => let h := Control.hyp h in print (of_constr h); apply $h
      end.

.. _ltac2_match_goal_hyps_rev_ex:

.. example:: Matching hypotheses with reverse

   (Equivalent to this :ref:`Ltac1 example<match_goal_hyps_rev_ex>`.)

   Hypotheses are matched from the first hypothesis to the last until the :tacn:`apply` succeeds.

   .. coqtop:: all abort

      Goal forall A B : Prop, A -> B -> (A->B).
      intros.
      match! reverse goal with
      | [ h : _ |- _ ] => let h := Control.hyp h in print (of_constr h); apply $h
      end.

.. _ltac2_match_goal_multiple_hyps_ex:

.. example:: Multiple ways to match a hypotheses

   (Equivalent to this :ref:`Ltac1 example<match_goal_multiple_hyps_ex>`.)

   Every possible match for the hypotheses is evaluated until the right-hand

>    side succeeds.  Note that `H1` and `H2` are never matched to the same hypothesis.

Reviewer: ```suggestion
   side succeeds.  Note that `h1` and `h2` are never matched to the same hypothesis.
```

   the name of a nonterminal describing the syntax (which the documentation calls :token:`term`).
   `print` and `of_constr` are functions provided by |Coq| through `Message.v`.

   .. todo "print" doesn't seem to pay attention to "Set Printing All"

   .. example:: Printing a :n:`@term`

      .. coqtop:: none

         Goal True.

      .. coqtop:: all

         From Ltac2 Require Import Message.
         Ltac2 Notation "ex1" x(constr) := print (of_constr x).
         ex1 (1 + 2).

   There are also metasyntactic classes described :ref:`here<syntactic_classes>`
   that combine other items.  For example, `list1(constr, ",")`
   recognizes a comma-separated list of one or more :token:`term`\s.

   .. example:: Parsing a list of :n:`@term`\s

      .. coqtop:: abort all

         Ltac2 rec print_list x := match x with
         | a :: t => print (of_constr a); print_list t
         | [] => ()
         end.
         Ltac2 Notation "ex2" x(list1(constr, ",")) := print_list x.
         ex2 1, 2, 3.

   An Ltac2 notation adds a parsing rule to the Ltac2 grammar, which is expanded
   to the provided body where every token from the notation is let-bound to the

>    corresponding generated expression.

Reviewer: Maybe add a note that, much like in Gallina, the RHS is *not* typechecked until substitution time?

      I don't when which is required

   NOT RIGHT YET: In the :token:`ltac2_match_pattern`\s, metavariables have the form :n:`?@ident`, whereas

>    in the :n:`@ltac2_expr`\s, the question mark is omitted.

Reviewer: This seems right now, no?

Author: Yeah.  If the right-hand side refers to a global symbol `x`, then it would be a user error to choose `x` as a metavariable.  And presumably if `x` is a bound variable in the RHS that will be handled correctly (not substituted).

   :n:`@cpattern`
      The syntax of :token:`cpattern` is
      the same as that of :token:`term`\s, but it can contain pattern matching
      metavariables in the form :n:`?@ident` and :n:`@?@ident`.  :g:`_` can be used to match
      irrelevant terms.

      .. todo more on @?@ident here: https://github.com/coq/coq/pull/12085#discussion_r467504046
      .. todo Example is broken :ref:`Example<ltac2_match_with_holes_ex>`.

      .. todo Didn't understand the following 2 paragraphs well enough to revise
         see https://github.com/coq/coq/pull/12103#discussion_r436297754 for a
         possible example

>       Unlike Ltac1, Ltac2 :n:?id metavariables only match closed terms.

Author: Indeed I do.

   :n:`|- @ltac2_match_pattern`
      A pattern to match with the current goal

   Note that unlike Ltac1, only lowercase identifiers are valid as Ltac2

>    bindings.  Ltac2 will report an error if a bound variables
>    that starts with an uppercase character.

Reviewer: "if a bound variables" is incorrect grammar.  Maybe you meant
```suggestion
   bindings.  Ltac2 will report an error if you try to use a bound variable
   that starts with an uppercase character.
```
?

Author: A typo or omission on my part.

In addition, there are metasyntactic operations for describing
more complex syntax, such as making an item optional or representing a list of items.
When parsing, each syntactic class expression returns a value that's bound to a name in the
notation definition.

Syntactic classes are described with a form of S-expression:

   .. insertprodn ltac2_scope ltac2_scope

   .. prodn::
      ltac2_scope ::= @string
      | @int
      | @name
      | @name ( {+, @ltac2_scope } )

.. todo no syn class for ints or strings?
   parm names are not reserved (e.g the var can be named "list1")

Metasyntactic operations that can be applied to other syntactic classes are:

  :n:`opt(@ltac2_scope)`
    Parses an optional :token:`ltac2_scope`.  The associated value is either :n:`None` or
    enclosed in :n:`Some`

  :n:`list1(@ltac2_scope {? , @string })`
    Parses a list of one or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
    items must be separated by :token:`string`.

  :n:`list0(@ltac2_scope {? , @string })`
    Parses a list of zero or more :token:`ltac2_scope`\s.  If :token:`string` is specified,

>     items must be separated by :token:`string`.  For zero items, the associated value

Reviewer: ```suggestion
    items must be separated by :token:`string`; otherwise, they must be whitespace-separated.  For zero items, the associated value
```

Author: That's true, but I don't think it's necessary to repeat standard lexing rules here.  We use lists all over the place in the grammar without repeating this rule.  So I would leave this as-is.

    is an empty list.

  :n:`seq({+, @ltac2_scope })`
    Parses the :token:`ltac2_scope`\s in order.  The associated value is a tuple,
    omitting :token:`ltac2_scope`\s that are :token:`string`\s.
    `self` and `next` are not permitted within `seq`.

The following classes represent nonterminals with some special handling.  The
table further down lists the classes that that are handled plainly.

  :n:`constr {? ( {+, @scope_key } ) }`
    Parses a :token:`term`.  If specified, the :token:`scope_key`\s are used to interpret
    the term (as described in  :ref:`LocalInterpretationRulesForNotations`).  The last
    :token:`scope_key` is the top of the scope stack that's applied to the :token:`term`.

    .. todo PR the original text describes "constr:(c%A%B)" which I couldn't make sense of
       Is that something we should keep (and if so, what is an example?)

  :n:`open_constr`
    Parses an open :token:`term`.

  :n:`ident`
    Parses :token:`ident` or :n:`$@ident`.  The first form returns :n:`ident:(@ident)`,
    while the latter form returns the variable :n:`@ident`.

  :n:`tactic {? (@int) }`

>     Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
>     parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
>     `tactic(6)` parses :token:`ltac2_expr`.
>     :token:`int` must be in the range `0 .. 6`.

Reviewer: ```suggestion
    Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
    `tactic(6)` parses :token:`ltac2_expr`.
    :token:`int` must be in the range `0 .. 6`.
    
    Note that the syntax for integers, strings, tuples, and lists is part of :token:`ltac2_expr`,
    so you can use `tactic` scope to parse these constructs.
```

This will fix #12805

   :name: Ltac2 Notation (abbreviation)

   Introduces a special kind of notation, called an abbreviation,
   that does not add any parsing rules. It is similar in
   spirit to Coq abbreviations (see :cmd:`Notation (abbreviation)`,
   insofar as its main purpose is to give an
   absolute name to a piece of pure syntax, which can be transparently referred to
   by this name as if it were a proper definition.

   The abbreviation can then be manipulated just like a normal Ltac2 definition,
   except that it is expanded at internalization time into the given expression.
   Furthermore, in order to make this kind of construction useful in practice in
   an effectful language such as Ltac2, any syntactic argument to an abbreviation
   is thunked on-the-fly during its expansion.

   For instance, suppose that we define the following.

   :n:`Ltac2 Notation foo := fun x => x ().`

   Then we have the following expansion at internalization time.

   :n:`foo 0 ↦ (fun x => x ()) (fun _ => 0)`

   Note that abbreviations are not type checked at all, and may result in typing
   errors after expansion.

.. _defining_tactics:

Defining tactics
~~~~~~~~~~~~~~~~

> All tactics used in Ltac2 must defined through notations.  These include built-in tactics (those

Reviewer: ```suggestion
All tactics used in Ltac2 must be defined through notations.  These include built-in tactics (those
```
This seems false, though, as you can define tactics with `Ltac2 foo := ...`.  Or, at least, it seems unclear what is meant to a newcomer.

Reviewer 2: Indeed, nothing forces you to go through notations and it's a good thing. It'd be terrible to have to...

Reviewer: I think the intended meaning was something like "you can't just call Ltac1 tactics in Ltac2 without defining notations for them or wrapping in ltac1:(...)"?

defined in OCaml code in the |Coq| executable) and Ltac1 tactics, which are defined in `.v` files.

Notations for many but not all built-in tactics are defined in `Notations.v`, which is automatically
loaded with Ltac2.  The Ltac2 syntax for these tactics is often identical or very similar to the
tactic syntax described in other chapters of this documentation.  These notations rely on tactic functions
declared in `Std.v`.  Functions corresponding to some built-in tactics may not yet be defined in the
|Coq| executable or declared in `Std.v`.  Adding them may require code changes to |Coq| or defining
workarounds through Ltac1 (described below).

Two examples of syntax differences:

- There is no notation defined that's equivalent to :n:`intros until {| @ident | @num }`.  There is,
  however, already an ``intros_until`` tactic function defined ``Std.v``, so it may be possible for a user
  to add the necessary notation.

> - The build-in `simpl` tactic supports the use of scope keys in delta flags, e.g. :n:`simpl ["+"%nat]`

Reviewer: ```suggestion
- The built-in `simpl` tactic in Ltac1 supports the use of scope keys in delta flags, e.g. :n:`simpl ["+"%nat]`
```

  which is not accepted by Ltac2.  This is because Ltac2 uses a different
  definition for :token:`delta_flag`; compare it to :token:`ltac2_delta_flag`.  This also affects
  :tacn:`compute`.

Ltac1 tactics are not automatically available in Ltac2.  (Note that some of the tactics described
in the documentation are defined with Ltac1.)

> You can make them accessible in Ltac2 with the following commands:

Reviewer: ```suggestion
You can make them accessible in Ltac2 with commands similar to the following:
```

.. coqtop:: in

   From Coq Require Import Lia.
   Local Ltac2 lia_ltac1 () := ltac1:(lia).
   Ltac2 Notation "lia" := lia_ltac1 ().

A similar approach can be used to access missing built-in tactics.  See :ref:`simple_api` for an
example that passes two parameters to a missing build-in tactic.

.. _syntactic_classes:

Syntactic classes
~~~~~~~~~~~~~~~~~

The simplest syntactic classes in Ltac2 notations represent individual nonterminals
from the |Coq| grammar.  Only a few selected nonterminals are available as syntactic classes.
In addition, there are metasyntactic operations for describing

> more complex syntax, such as making an item optional or representing a list of items.

Reviewer: This seems to duplicate what was said above, at least a little bit; is this intentional?

Author: Can you be more specific?  Hard to guess your meaning.

Reviewer: Ah, nevermind, the thing I thought it was duplicative with was actually just a short bit of text that referred to this section.

When parsing, each syntactic class expression returns a value that's bound to a name in the
notation definition.

Syntactic classes are described with a form of S-expression:

   .. insertprodn ltac2_scope ltac2_scope

   .. prodn::
      ltac2_scope ::= @string
      | @int
      | @name
      | @name ( {+, @ltac2_scope } )

> .. todo no syn class for ints or strings?

Reviewer: Wouldn't that fall under `@name`?  e.g., `integer`, `string` are both `@name`s, no?

Author: Nope and nope.

```
msg constr:(True).     (* parses *)
msg string:("abc").     (* syntax error on the colon *)
```

Reviewer: I think you're confusing `ltac2_scope` with `ltac2_quotations`.  This is about things like the `constr` in
```coq
Ltac2 Notation "foo" c(constr) := c.
```
Incidentally, there don't seem to be scopes for int and string, and I believe this may be an oversight.  I've reported this as https://github.com/coq/coq/issues/12805

Author: > I think you're confusing `ltac2_scope` with `ltac2_quotations`.

Maybe so.  After a month's delay a lot of the detail fades.  Probably I just cooked up the wrong example today.  OTOH, if you have syntax class `foo` why wouldn't you have a matching built-in quotation `foo:(...)` (with a few obvious exceptions, e.g. no list1).

Reviewer: I believe the reason is that `ltac2_scope` is intended to eventually be user extensible, while there's no need to hard-code extensibility of `ltac2_quotations` because it can already be extended, in some sense, via `Ltac2 Notation`

>   parm names are not reserved (e.g the var can be named "list1"

Reviewer: ```suggestion
  parm names are not reserved (e.g the var can be named "list1")
```

Author: Thanks.  FWIW, that's a non-user visible comment.

Metasyntactic operations that can be applied to other syntactic classes are:

  :n:`opt(@ltac2_scope)`
    Parses an optional :token:`ltac2_scope`.  The associated value is either :n:`None` or
    enclosed in :n:`Some`

  :n:`list1(@ltac2_scope {? , @string })`
    Parses a list of one or more :token:`ltac2_scope`\s.  If :token:`string` is specified,

>     items must be separated by :token:`string`.

Reviewer: What are the accepted separator(s) if no string is provided?  Only spaces?

Reviewer 2: Yep.

Reviewer: ```suggestion
    items must be separated by :token:`string`.  If no :token:`string` is given,
    the default is whitespace.
```

Author: Not necessary to repeat basic lexing rules here.  We wouldn't want to say "separated by :token:`string`\s with white space before and after :token:`string` if necessary", either.

  :n:`list0(@ltac2_scope {? , @string })`
    Parses a list of zero or more :token:`ltac2_scope`\s.  If :token:`string` is specified,
    items must be separated by :token:`string`.  For zero items, the associated value
    is an empty list.

  :n:`seq({+, @ltac2_scope })`
    Parses the :token:`ltac2_scope`\s in order.  The associated value is a tuple,
    omitting :token:`ltac2_scope`\s that are :token:`string`\s.
    `self` and `next` are not permitted within `seq`.

The following classes represent nonterminals with some special handling.  The
table further down lists the classes that that are handled plainly.

  :n:`constr {? ( @scope_key ) }`
    Parses a :token:`term`.  If specified, the :token:`scope_key` is used to interpret
    the term (as described in  :ref:`LocalInterpretationRulesForNotations`)

    .. todo PR the original text describes "constr:(c%A%B)" which I couldn't make sense of

>        Is that something we should keep (and if so, what is an example?)

Reviewer: Yes, we should keep it.  An example is
```coq
Require Import Ltac2.Ltac2.

Ltac2 Notation "foo" c(constr(nat)) := Message.print (Message.of_constr c).
Ltac2 Notation "bar" c(constr(type)) := Message.print (Message.of_constr c).

Goal True.
  foo (1+1).
  Fail bar (1+1). (* The term "1" has type "nat" while it is expected to have type "Type". *)
  bar (nat+nat). (* (nat + nat)%type *)
  Fail foo (nat+nat). (* The term "nat" has type "Set" while it is expected to have type "nat". *)
```
An example of nested scopes is
```coq
Require Import Ltac2.Ltac2.
Require Import Coq.Strings.String.
Require Import Coq.Lists.List.

Ltac2 Notation "foo" c(constr(string,list)) := Message.print (Message.of_constr c).

Goal True.
  foo "a".
  foo ("a" ++ "a").
  foo (1 :: nil).
  Fail foo (@nil nat ++ nil).
  foo (@nil nat ++ nil)%list.
```
where doing `constr(A,B)` is equivalent to interpreting in the scope stack `%A%B`.  (Note that your syntax is wrong, it allows a comma-delimited list.)

Author: > (Note that your syntax is wrong, it allows a comma-delimited list.)

I corrected this, but I don't think an example is necessary.  Not easy to get the syntax since this isn't in the mlg files but only in OCaml code that I don't fully understand:

```
let () = add_scope "constr" (fun arg ->
    let delimiters = List.map (function
        | SexprRec (_, { v = Some s }, []) -> s
        | _ -> scope_fail "constr" arg)
        arg
    in
    let act e = Tac2quote.of_constr ~delimiters e in
    Tac2entries.ScopeRule (Pcoq.Symbol.nterm Pcoq.Constr.constr, act)
  )
```

>   :n:`open_constr`

Reviewer: I think it's a bug that `open_constr` does not permit scopes like `constr`, want to report it?

Author: I will leave that to you.

Reviewer: https://github.com/coq/coq/issues/12806

    Parses an open :token:`term`.

  :n:`ident`
    Parses :token:`ident` or :n:`$@ident`.  The first form returns :n:`ident:(@ident)`,
    while the latter form returns the variable :n:`@ident`.

  :n:`tactic {? (@int) }`
    Parses an :token:`ltac2_expr`.  If :token:`int` is specified, the construct
    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
    `tactic(6)` parses :token:`ltac2_expr`.
    :token:`int` must be in the range `0 .. 6`.

    .. todo this doesn't work as expected: "::" is in ltac2_expr1
       Ltac2 Notation "ex4" x(tactic(0)) := x.
       ex4 auto :: [auto].

  .. not sure "self" and "next" do anything special.  I get the same error
     message for both from constructs like

     Ltac2 Notation "ex5" x(self) := auto.
     ex5 match.

>      Syntax error: [tactic:tac2expr level 5] expected after 'match' (in [tactic:tac2expr]).

Reviewer: Is there a bug report for this?  cc @ppedrot

  :n:`self`
    parses an Ltac2 expression at the current level and returns it as is.

  :n:`next`

>     parses an Ltac2 expression at the next level and returns it as is.

Reviewer: Next level up, or next level down?  (And which is tighter?)

Reviewer: (I think it's next level down, and tighter, but I'm not sure, and in any case this should be documented, either here or in the Gallina notations section (if it's not already) or both)

Author: I think these are equivalent to the parser's SELF and NEXT but now looking at it again I don't see how that makes any sense.  @ppedrot can you give an example showing how to use it?

Reviewer 2: It's the direct equivalent of camlp5 NEXT / SELF, so it makes sense w.r.t. the current tactic level being expanded. Not sure whether I have a practical example at hand though.

Reviewer: Hmmm, this feature seems to be sort-of broken anyway, see https://github.com/coq/coq/issues/11866 and https://github.com/coq/coq/issues/12807 and https://github.com/coq/coq/issues/12808

Author: SELF and NEXT only make sense in the context of defining productions for some nonterminal.  Notations with precedence number would already substitute SELF/NEXT for the variables in the expression.  Seems like this may have been added without thinking it through, especially if you can't provide an example.

  :n:`thunk(@ltac2_scope)`
    Used for semantic effect only, parses the same as :token:`ltac2_scope`.
    If :n:`e` is the parsed expression for :token:`ltac2_scope`, `thunk`
    returns :n:`fun () => e`.

  :n:`pattern`
    parses a :token:`cpattern`

  .. todo PR I dropped "terminal" and "keyword" because they don't seem to behave
     any differently from just giving a string, e.g. for the keyword "match",

>      x(terminal("match")) worked the same as x("match")

Reviewer: cc @herbelin , maybe, can you (or someone else) check if these do different things?

    parses a :n:`ltac2_expr@int`, for example `tactic(5)` parses :token:`ltac2_expr5`.
    `tactic(6)` parses :token:`ltac2_expr`.
    :token:`int` must be in the range `0 .. 6`.

    You can also use `tactic` to accept an :token:`int` or a :token:`string`, but there's
    no syntactic class that accepts *only* an :token:`int` or a :token:`string`.

    .. todo this doesn't work as expected: "::" is in ltac2_expr1
       Ltac2 Notation "ex4" x(tactic(0)) := x.
       ex4 auto :: [auto].

  .. not sure "self" and "next" do anything special.  I get the same error
     message for both from constructs like

     Ltac2 Notation "ex5" x(self) := auto.
     ex5 match.

     Syntax error: [tactic:tac2expr level 5] expected after 'match' (in [tactic:tac2expr]).

  :n:`self`
    parses an Ltac2 expression at the current level and returns it as is.

  :n:`next`
    parses an Ltac2 expression at the next level and returns it as is.

  :n:`thunk(@ltac2_scope)`
    Used for semantic effect only, parses the same as :token:`ltac2_scope`.
    If :n:`e` is the parsed expression for :token:`ltac2_scope`, `thunk`
    returns :n:`fun () => e`.

  :n:`pattern`
    parses a :token:`cpattern`

>   .. todo PR I dropped "terminal" and "keyword" because they don't seem to behave
>      any differently from just giving a string, e.g. for the keyword "match",
>      x(terminal("match")) worked the same as x("match")

Reviewer: ```suggestion
```
You added them back.  (Feel free to ignore this suggestion, though, as it's only about a comment)

Author: Pierre-Marie thought I should have them, I'll drop the note.
